### PR TITLE
Reads Set Editor cell update

### PR DESCRIPTION
**Pull request by eapearson**
These changes are mostly focused on the grouped params for the reads set and data flow.
Next is usability, including initial editor experience, persistence of editor selection, editor compactness.

Looks like a lot of changes -- and it was a lot of work -- but most of them are due to forking the app cell widgets and utilities to use in the editor cell.

Comments for this PR:

- the flow currently automatically selects the first available reads set to populate the editor

- other than that it is:

add editor

select a reads set to edit or create a new one with a given name

at present you may select a new reads set to edit at any time, but selecting a reads set could also lock down the editor to just that one (making the selection area disappear), with the option of unlocking the selection area to select or create another.

the reads set name _should_ allow renaming, but does not now; at present if you change the name it allows you to create a new reads set with the new name (leaving the old one intact.) I don't think this is necessarily very natural behavior.

It _should_ also persist selecte reads set and restore that when the narrative is reloaded -- it does not currently do that.

recent changes:

- fork off the editor controls from the app cell
(we can make improvements locally withiout affecting the app cells, but should update app cells
to use these new controls soon - then we can collapse the fork (appWidgets2).)

- "touched" state for controls which flow up to the editor to allow for catching the time between a control being edited and the update ("change" event) propagating control data up to the editor.

- Crude status panel at the bottom for feedback

- debugging artifacts include the yellow highlight when touched state is active, the param flags at the bottom

- "struct" control type -- arbitrary collection of named controls (aka grouped parameters)

- set (list) of struct controls (this is how the items parameter is implemented)

- arbitrary nesting of controls - for allowing sets of grouped params (which may themselves contain sets of grouped params, etc.)

- unused new controls (because I thought we needed them but don't: map control (key-value), data attachments (key-ref)

- driven by a "spec" inside editObjectSelector, which is not the same as but derived from the app spec. Allows re-using existing controls, and composable controls.

incomplete or issues

- reads set name: should it rename the current set or create a duplicate?
- if rename, should we have a Duplicate button?

- deletion from within the editor (just need a delete button and simple implementation)

- make controls more compact

- move help (i button) to a pop up, if at all

- required constraint no longer creates red arrow - fix

- item uniqueness and select only having unused objects not enforced (yet - it was on the old controls but not the new)

- need new control for object selection (separate task). Other than workspace-searchable, it also needs to have richer information and to "stick" the with the current value.

For instance, if an object is selected, but is later updated, a dropdown based on the current workspace offerings will not show the currently selected object, since it is an old version. With object-name based values, this is not a problem since they imply the most recent version. With references though we are faced with having a set of possible new objects to select from plus the existing reference to leave in place.

Different ways of handling this...

- a few more items need to be extracted to make this re-usable for different editors

- and i'm sure lots more ...

### CODE REVIEW

**Comment by eapearson** [comment]
Oh, yes, to test it you you need to use the "Test Create App" from NarrativeTest.

**Comment by briehl** [comment]
So far so good!
Except the editor isn't appearing for me. I see this in the console instead:
```
starting model...          readsSetModel.js:185 
GOT IT! Object             readsSetModel.js:189 
sending ready...           readsSetUpdateEditor.js:388
Unhandled rejection TypeError: Cannot read property 'make' of undefined
    at makeSingleInputControl (http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:163:48)
    at http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:325:24
    at Array.map (native)
    at makeInputControl (http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:324:37)
    at render (http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:348:27)
    at http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:65:21
    at r (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:31:5015)
    at r._settlePromiseFromHandler (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:30:10865)
    at r._settlePromise (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:30:11672)
    at r._settlePromiseCtx (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:30:12180)
    at n._drainQueue (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:29:2991)
    at n._drainQueues (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:29:3069)
    at drainQueues (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:29:1209)O @ bluebird.min.js:29
bluebird.min.js:29 Unhandled rejection TypeError: Cannot read property 'make' of undefined
    at makeSingleInputControl (http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:163:48)
    at http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:325:24
    at Array.map (native)
    at makeInputControl (http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:324:37)
    at render (http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:348:27)
    at http://localhost:8889/narrative/static/kbase/js/widgets/appWidgets2/input/multiStructInput.js:65:21
    at r (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:31:5015)
    at r._settlePromiseFromHandler (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:30:10865)
    at r._settlePromise (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:30:11672)
    at r._settlePromiseCtx (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:30:12180)
    at n._drainQueue (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:29:2991)
    at n._drainQueues (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:29:3069)
    at drainQueues (http://localhost:8889/narrative/static/ext_components/bluebird/js/browser/bluebird.min.js:29:1209)
```

**Comment by eapearson** [comment]
okay, i'll do a fresh pull; maybe something wasn't captured in the commits.

**Comment by eapearson** [comment]
A fresh build from PR#815 works for me. Were you using a fresh narrative, or one with an existing editor cell?

**Comment by briehl** [comment]
I was using a new editor cell, but existing objects. I'll try from scratch.

**Comment by briehl** [comment]
But I did it from a fresh install of your branch... is your branch ahead of the PR?

**Comment by briehl** [comment]
Now I see a more obvious error - 
```
Uncaught SyntaxError: Unexpected token )    structInput.js:254
```

**Comment by eapearson** [comment]
I pushed one change after the PR, but the PR picked that up (appmanager.py mysteriously developed a merge conflict).

**Comment by eapearson** [comment]
okay let me switch browsers too in case something is cached...

**Comment by eapearson** [comment]
yeah, i see that.

**Comment by eapearson** [comment]
my editor has gone wonky ...

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/structInput.js [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+                    console.error(err);
+                    ui.setContent('input-container', 'ERROR!' + err.message);
+                });
+        }
+        
+        function render(events) {
+            container.innerHTML = div({
+                        dataElement: 'main-panel'
+                    }, [
+                        div({dataElement: 'input-container'}, )
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/structInput.js]: Here's the error - the extra trailing comma needs to go.

Comment (eapearson) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/structInput.js]: yes, i saw that. i guess safari is more lenient...

**Comment by eapearson** [comment]
Okay, that change is pushed up now. Weird, I still could not replicate it.

**Comment by briehl** [comment]
Chrome really didn't like it very much. :\ I'll pull and try again.

**Comment by briehl** [comment]
It works! I agree with your notes and caveats, but it works. Let's merge it up.

**Pull request merged**